src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,416 @@
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.BraceCompletion
+{
+    [Export(LanguageNames.CSharp, typeof(IBraceCompletionService)), Shared]
+    internal class CurlyBraceCompletionService : AbstractBraceCompletionService
+    {
+        private static readonly SyntaxAnnotation s_closingBraceSyntaxAnnotation = new("original closing brace");
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CurlyBraceCompletionService()
+        {
+        }
+
+        protected override char OpeningBrace => CurlyBrace.OpenCharacter;
+
+        protected override char ClosingBrace => CurlyBrace.CloseCharacter;
+
+        public override async Task<bool> AllowOverTypeAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            return await CheckCurrentPositionAsync(context.Document, context.CaretLocation, cancellationToken).ConfigureAwait(false)
+                && await CheckClosingTokenKindAsync(context.Document, context.ClosingPoint, cancellationToken).ConfigureAwait(false);
+        }
+
+        public override async Task<BraceCompletionResult?> GetTextChangesAfterCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+        {
+            // Format the span from the open brace location up to and including the closing brace location.
+            var (formattingChanges, newClosingPoint) = await FormatTrackingSpanAsync(braceCompletionContext.Document, braceCompletionContext.OpeningPoint, braceCompletionContext.ClosingPoint,
+                shouldHonorAutoFormattingOnCloseBraceOption: true, rules: null, cancellationToken).ConfigureAwait(false);
+
+            if (formattingChanges.IsEmpty)
+            {
+                return null;
+            }
+
+            // The caret location should be at the start of the closing brace character.
+            return new BraceCompletionResult(formattingChanges, newClosingPoint - 1);
+        }
+
+        public override async Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext context, bool supportsVirtualSpace, CancellationToken cancellationToken)
+        {
+            var document = context.Document;
+            var closingPoint = context.ClosingPoint;
+            var openingPoint = context.OpeningPoint;
+            var documentSnapshotText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            // Create a new source text that is not based on an editor snapshot so we can take advantage
+            // of the multi-version text change merging in ChangedText instead of ChangedSnapshotText.
+            var originalDocumentText = SourceText.From(documentSnapshotText.ToString(), documentSnapshotText.Encoding, documentSnapshotText.ChecksumAlgorithm);
```

Comment:
=-o...

i am... really not loving this.  esp as that .ToString and SourceText.From could be very expensive.

Reply:
i presume you're using this to get access to the 'Merge' helper in that type.  But it feels to me like that could be extracted out and we could operate on the low level pieces of data (i.e. the TextChangesRanges) directly, accessing tha thelper to get the final result.  is that not hte case?

Reply:
Yeah essentially - I can do that, but if you don't mind I'll do that after the rest of the PR looks good as I'll need a compiler review on the change in ChangedText.  I'd rather not spam them too much yet :)